lib/expand/join.go

```diff
@@ -150,18 +150,23 @@ func (p *Peer) Execute(req *installpb.ExecuteRequest, stream installpb.Agent_Exe
 		case event := <-p.dispatcher.Chan():
 			err := stream.Send(event)
 			if err != nil {
-				return trace.Wrap(err)
+				return err
 			}
 		case req := <-p.closeC:
 			err := stream.Send(req.resp)
 			close(req.doneC)
 			if err != nil {
-				return trace.Wrap(err)
+				return err
 			}
 		case err := <-p.execDoneC:
 			if err != nil {
-				// Phase finished with an error
-				return status.Error(codes.FailedPrecondition, err.Error())
+				// Phase finished with an error.
+				if installpb.IsRPCError(err) {
+					return trace.Unwrap(err)
+				}
+				// Aborted code is RPC-specific here - it indicates an operation
+				// failure that can be retried
+				return status.Error(codes.Aborted, err.Error())
```

Comment:
I've been previously using `codes.FailedCondition` as the error type for execute failures but I reconsidered it after re-reading the comment:
```go
// FailedPrecondition indicates operation was rejected because the
// system is not in a state required for the operation's execution.
// For example, directory to be deleted may be non-empty, an rmdir
// operation is applied to a non-directory, etc.
//
// A litmus test that may help a service implementor in deciding
// between FailedPrecondition, Aborted, and Unavailable:
//  (a) Use Unavailable if the client can retry just the failing call.
//  (b) Use Aborted if the client should retry at a higher-level
//      (e.g., restarting a read-modify-write sequence).
//  (c) Use FailedPrecondition if the client should not retry until
//      the system state has been explicitly fixed. E.g., if an "rmdir"
//      fails because the directory is non-empty, FailedPrecondition
//      should be returned since the client should not retry unless
//      they have first fixed up the directory by deleting files from it.
//  (d) Use FailedPrecondition if the client performs conditional
//      REST Get/Update/Delete on a resource and the resource on the
//      server does not match the condition. E.g., conflicting
//      read-modify-write on the same resource.
```
which made me believe `codes.Aborted` is better suited here and I'm using `codes.FailedPrecondition` to mark the early stage initialization failures.

Reply:
Nit: This answer should probably be summarized in a comment, to answer the why codes.Aborted...